classes/observers/traits/UsageEvent.inc.php

```diff
@@ -0,0 +1,164 @@
+                        $canonicalUrlParams[] = $publicationId;
+                    }
+                }
+                $canonicalUrlParams[] = $representation->getId();
+                $canonicalUrlParams[] = $submissionFile->getId();
+                break;
+            case Application::ASSOC_TYPE_SUBMISSION:
+                $canonicalUrlOp = 'view';
+                if ($application->getName() == 'omp') {
+                    $canonicalUrlOp = 'book';
+                }
+                $canonicalUrlParams = [$submission->getId()];
+                $router = $this->request->getRouter(); /** @var PageRouter $router */
+                $op = $router->getRequestedOp($this->request);
+                $args = $router->getRequestedArgs($this->request);
+                if ($op == $canonicalUrlOp && count($args) > 1) {
+                    if ($args[1] == 'version' && count($args) == 3) {
+                        $publicationId = (int) $args[2];
+                        $canonicalUrlParams[] = 'version';
+                        $canonicalUrlParams[] = $publicationId;
+                    }
+                }
+                break;
+            case Application::getContextAssocType():
+                $canonicalUrlOp = '';
+                $canonicalUrlPage = 'index';
+                break;
+        }
+        $canonicalUrl = $this->getCanonicalUrl($this->request, $canonicalUrlPage, $canonicalUrlOp, $canonicalUrlParams);
+
+        $this->time = Core::getCurrentDate();
+        $this->assocType = $assocType;
+        $this->canonicalUrl = $canonicalUrl;
+        $this->context = $context;
+        $this->submission = $submission;
+        $this->representation = $representation;
+        $this->submissionFile = $submissionFile;
+        $this->version = Registry::get('appVersion');
+    }
+
+    /**
+     * Get the canonical URL for the usage object
+     */
+    protected function getCanonicalUrl(Request $request, string $canonicalUrlPage = null, string $canonicalUrlOp = null, array $canonicalUrlParams = null): string
+    {
+        $router = $request->getRouter(); /** @var PageRouter $router */
+        $context = $router->getContext($request);
+
+        if (!isset($canonicalUrlPage)) {
+            $canonicalUrlPage = $router->getRequestedPage($request);
+        }
+        if (!isset($canonicalUrlOp)) {
+            $canonicalUrlOp = $router->getRequestedOp($request);
+        }
+        if (!isset($canonicalUrlParams)) {
+            $canonicalUrlParams = $router->getRequestedArgs($request);
+        }
+
+        $canonicalUrl = $router->url(
+            $request,
+            null,
+            $canonicalUrlPage,
+            $canonicalUrlOp,
+            $canonicalUrlParams
+        );
+
+        // Make sure we log the server name and not aliases.
+        $configBaseUrl = Config::getVar('general', 'base_url');
+        $requestBaseUrl = $request->getBaseUrl();
+        if ($requestBaseUrl !== $configBaseUrl) {
```

Comment:
This is adapted from elsewhere in the codebase, right? This kind of URL logic gives me the heebie-jeebies and I'll be much happier when we leave to a framework!